snyk.go

```diff
@@ -1,24 +1,77 @@
+// ProjectsFilter is the top level filter type of filtering Snyk project
+type ProjectsFilter struct {
+	Filters ProjectsFilterBody `json:"filters"`
+}
+
+type ProjectsFilterBody struct {
+	Attributes ProjectsFiltersAttributes `json:"attributes"`
+}
+
+type ProjectsFiltersAttributes struct {
+	Criticality []string `json:"criticality,omitempty"`
+	Environment []string `json:"environment,omitempty"`
+	Lifecycle   []string `json:"lifecycle,omitempty"`
+}
+
+func getOrgProjects(flags flags, customDebug debug) (jsn.Json, error) {
+	// According to https://snyk.docs.apiary.io/#reference/projects/all-projects/list-all-projects this should be
+	// a POST API call but historically we used GET here. The following code maintains backwards compatibility for
+	// existing cases where people aren't filtering projects by attributes, as it appears the API does not return
+	// the full project list with empty attribute filters in the request body.
+	verb := "GET"
+	body := ProjectsFilter{}
+
+	if len(flags.optionalFlags.projectCriticality) > 0 {
+		body.Filters.Attributes.Criticality = strings.Split(flags.optionalFlags.projectCriticality, ",")
+		verb = "POST"
+	}
+
+	if len(flags.optionalFlags.projectEnvironment) > 0 {
+		body.Filters.Attributes.Environment = strings.Split(flags.optionalFlags.projectEnvironment, ",")
+		verb = "POST"
+	}
+
+	if len(flags.optionalFlags.projectLifecycle) > 0 {
+		body.Filters.Attributes.Lifecycle = strings.Split(flags.optionalFlags.projectLifecycle, ",")
+		verb = "POST"
+	}
+
+	var marshalledBody []byte
+	var err error
+
+	if verb == "POST" {
+		marshalledBody, err = json.Marshal(body)
+		if err != nil {
+			log.Printf("*** ERROR *** Not able to read attributes %s\n", flags.mandatoryFlags.endpointAPI)
+			errorMessage := "Failure, Not able to read attributes\n"
+			err = errors.New(errorMessage)
+			verb = "GET"
+			marshalledBody = nil
+		}
+	}
+
+	responseData, err := makeSnykAPIRequest(verb, flags.mandatoryFlags.endpointAPI+"/v1/org/"+flags.mandatoryFlags.orgID+"/projects", flags.mandatoryFlags.apiToken, marshalledBody, customDebug)
 	if err != nil {
-		log.Printf("*** ERROR *** Could not get the Project(s) for endpoint %s\n", Mf.endpointAPI)
-		errorMessage := "Failure, Could not get the Project(s) for endpoint " + Mf.endpointAPI + "\n"
+		log.Printf("*** ERROR *** Could not get the Project(s) for endpoint %s\n", flags.mandatoryFlags.endpointAPI)
+		errorMessage := "Failure, Could not get the Project(s) for endpoint " + flags.mandatoryFlags.endpointAPI + "\n"
 		err = errors.New(errorMessage)
 	}
 
 	project, err := jsn.NewJson(responseData)
 	if err != nil {
-		log.Printf("*** ERROR *** Could not get read the response from endpoint %s\n", Mf.endpointAPI)
-		errorMessage := "Failure, Could not get read the response from endpoint " + Mf.endpointAPI + "\n"
+		log.Printf("*** ERROR *** Could not get read the response from endpoint %s\n", flags.mandatoryFlags.endpointAPI)
```

Comment:
Rephrase: `*** ERROR *** Could not read the response from endpoint %s\n", flags.mandatoryFlags.endpointAPI)`